File: live/player/resolve_cache.py
"""解析结果短期缓存：房间 meta 与各档位 tier 分层。"""

from __future__ import annotations

import copy
import time
from threading import Lock

DEFAULT_TTL = 45.0

_lock = Lock()
_entries: dict[str, dict] = {}


def _now() -> float:
    return time.monotonic()


def _meta_key(site: str, room_id: str) -> str:
    return f"meta:{site}:{room_id}"


def _tier_key(site: str, room_id: str, quality_name: str) -> str:
    return f"tier:{site}:{room_id}:{quality_name}"


def _payload_key(site: str, room_id: str, mode: str, quality_key: str) -> str:
    return f"payload:{site}:{room_id}:{mode}:{quality_key}"


def get(key: str) -> dict | None:
    with _lock:
        entry = _entries.get(key)
        if not entry or _now() >= entry["expires"]:
            if entry:
                del _entries[key]
            return None
        return copy.deepcopy(entry["data"])


def set(key: str, data: dict, *, ttl: float = DEFAULT_TTL) -> None:
    with _lock:
        _entries[key] = {"data": copy.deepcopy(data), "expires": _now() + ttl}


def get_meta(site: str, room_id: str) -> dict | None:
    return get(_meta_key(site, room_id))


def set_meta(site: str, room_id: str, meta: dict, *, ttl: float = DEFAULT_TTL) -> None:
    set(_meta_key(site, room_id), meta, ttl=ttl)


def get_tier(site: str, room_id: str, quality_name: str) -> dict | None:
    return get(_tier_key(site, room_id, quality_name))


def set_tier(site: str, room_id: str, quality_name: str, tier: dict, *, ttl: float = DEFAULT_TTL) -> None:
    set(_tier_key(site, room_id, quality_name), tier, ttl=ttl)


def get_payload(site: str, room_id: str, mode: str, quality_key: str) -> dict | None:
    return get(_payload_key(site, room_id, mode, quality_key))


def set_payload(
    site: str,
    room_id: str,
    mode: str,
    quality_key: str,
    payload: dict,
    *,
    ttl: float = DEFAULT_TTL,
) -> None:
    set(_payload_key(site, room_id, mode, quality_key), payload, ttl=ttl)

File: live/node-server/src/cache/resolve-cache.ts
export const PAYLOAD_TTL = 60;
export const TIER_TTL = 60;
export const META_TTL = 180;
export const DEFAULT_TTL = PAYLOAD_TTL;
export const MAX_ENTRIES = 100;

interface CacheEntry {
  data: unknown;
  expires: number;
}

function now(): number {
  return performance.now() / 1000;
}

function metaKey(site: string, roomId: string): string {
  return `meta:${site}:${roomId}`;
}

function tierKey(site: string, roomId: string, qualityName: string): string {
  return `tier:${site}:${roomId}:${qualityName}`;
}

function payloadKey(site: string, roomId: string, mode: string, qualityKey: string): string {
  return `payload:${site}:${roomId}:${mode}:${qualityKey}`;
}

export class ResolveCache {
  private entries = new Map<string, CacheEntry>();
  private order: string[] = [];

  private purgeExpired(): void {
    const t = now();
    for (const key of [...this.order]) {
      const entry = this.entries.get(key);
      if (!entry || t >= entry.expires) {
        this.entries.delete(key);
        this.order = this.order.filter((k) => k !== key);
      }
    }
  }

  private evictIfNeeded(): void {
    while (this.order.length > MAX_ENTRIES) {
      const oldest = this.order.shift();
      if (oldest) {
        this.entries.delete(oldest);
      }
    }
  }

  get(key: string): unknown | null {
    this.purgeExpired();
    const entry = this.entries.get(key);
    if (!entry) {
      return null;
    }
    if (now() >= entry.expires) {
      this.entries.delete(key);
      this.order = this.order.filter((k) => k !== key);
      return null;
    }
    this.order = this.order.filter((k) => k !== key);
    this.order.push(key);
    return structuredClone(entry.data);
  }

  set(key: string, data: unknown, opts?: { ttl?: number }): void {
    const ttl = opts?.ttl ?? DEFAULT_TTL;
    this.purgeExpired();
    this.entries.delete(key);
    this.order = this.order.filter((k) => k !== key);
    this.entries.set(key, { data: structuredClone(data), expires: now() + ttl });
    this.order.push(key);
    this.evictIfNeeded();
  }

  stats(): {
    entries: number;
    max_entries: number;
    ttl_sec: { meta: number; tier: number; payload: number };
  } {
    this.purgeExpired();
    return {
      entries: this.entries.size,
      max_entries: MAX_ENTRIES,
      ttl_sec: {
        meta: META_TTL,
        tier: TIER_TTL,
        payload: PAYLOAD_TTL,
      },
    };
  }

  getMeta(site: string, roomId: string): Record<string, unknown> | null {
    const data = this.get(metaKey(site, roomId));
    return data && typeof data === "object" ? (data as Record<string, unknown>) : null;
  }

  setMeta(site: string, roomId: string, meta: Record<string, unknown>): void {
    this.set(metaKey(site, roomId), meta, { ttl: META_TTL });
  }

  getTier(site: string, roomId: string, qualityName: string): Record<string, unknown> | null {
    const data = this.get(tierKey(site, roomId, qualityName));
    return data && typeof data === "object" ? (data as Record<string, unknown>) : null;
  }

  setTier(
    site: string,
    roomId: string,
    qualityName: string,
    tier: Record<string, unknown>,
    opts?: { ttl?: number },
  ): void {
    this.set(tierKey(site, roomId, qualityName), tier, { ttl: opts?.ttl ?? TIER_TTL });
  }

  getPayload(
    site: string,
    roomId: string,
    mode: string,
    qualityKey: string,
  ): Record<string, unknown> | null {
    const data = this.get(payloadKey(site, roomId, mode, qualityKey));
    return data && typeof data === "object" ? (data as Record<string, unknown>) : null;
  }

  setPayload(
    site: string,
    roomId: string,
    mode: string,
    qualityKey: string,
    payload: Record<string, unknown>,
  ): void {
    this.set(payloadKey(site, roomId, mode, qualityKey), payload, { ttl: PAYLOAD_TTL });
  }
}
